Add new staging module - `constants`

dims opened the pull request:
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it

Introduces a new zero-dependency staging module `k8s.io/constants` that consolidates well-known Kubernetes constants in one foundational location, and wires apimachinery / api / kubectl to consume it as the single source of truth.

Layout:
- `rfc/`         — length constants for Kubernetes name-like fields (DNS-1123 label/subdomain, DNS-1035 label); docs describe the Kubernetes-specific grammars, not the RFCs.
- `limits/`      — `LabelValueMaxLength`, `LabelKeyMaxLength`, `FieldManagerMaxLength`.
- `labels/`      — well-known label keys (identifiers omit the `Label` prefix: `labels.Hostname`, `labels.TopologyZone`, …).
- `annotations/` — well-known annotation keys (identifiers omit the `Annotation`/`AnnotationKey` prefix/suffix).
- `taints/`      — well-known node taints (identifiers omit the `Taint` prefix).

#### Which issue(s) this PR is related to

Fixes #127889

#### Does this PR introduce a user-facing change?

```release-note
Added a new zero-dependency staging module k8s.io/constants containing well-known label, annotation, taint, RFC length, and Kubernetes-specific size constants. The existing constants in apimachinery and api/core/v1 are re-exported from the new module for backwards compatibility.
```

#### Additional documentation

- Issue: https://github.com/kubernetes/kubernetes/issues/127889

danwinship commented on staging/src/k8s.io/constants/annotations/annotations.go:
```diff
@@ -0,0 +1,105 @@
+	// AnnotationLoadBalancerSourceRangesKey is the key of the annotation on a
+	// service to set allowed ingress ranges on their LoadBalancers.
+	// It should be a comma-separated list of CIDRs.
+	AnnotationLoadBalancerSourceRangesKey = "service.beta.kubernetes.io/load-balancer-source-ranges"
```
This is deprecated; everyone should be using `service.spec.LoadBalancerSourceRanges` instead.

dims replied:
Fixed.

danwinship commented on staging/src/k8s.io/constants/labels/labels.go:
```diff
@@ -0,0 +1,91 @@
+// Well-known service labels.
+const (
+	// IsHeadlessService is added by Controller to an Endpoint denoting if its
+	// parent Service is Headless.
```
"... is added by the Endpoints and EndpointSlice controllers to Endpoints and EndpointSlices ..."

dims replied:
Applied the suggested wording

danwinship commented on staging/src/k8s.io/constants/labels/labels.go:
```diff
@@ -0,0 +1,91 @@
+	IsHeadlessService = "service.kubernetes.io/headless"
+)
+
+// Deprecated labels - kept for backwards compatibility.
```
inconsistent organization as compared to annotations/

dims replied:
Added a trailing // Deprecated labels block mirroring the annotations/ trailing Deprecated block; both use the Deprecated* naming convention.

danwinship commented on staging/src/k8s.io/constants/taints/taints.go:
```diff
@@ -0,0 +1,56 @@
+const (
+	// TaintNodeNotReady is added when node is not ready and removed when
+	// node becomes ready.
+	TaintNodeNotReady = "node.kubernetes.io/not-ready"
```
(naming redundancy again)

dims replied:
Done; api/core/v1/well_known_taints.go keeps the old TaintNode* names on the LHS.

danwinship commented:
@thockin and I were talking about something like this a while back (I had wanted `netutilsv2.IPv4` and `v1.IPv4Protocol` to be `==` to each other without casting, which is only possible if they can both depend on a typed `"IPv4"` string from some package which both `k/utils` and `k/api` would be allowed to depend on.) He suggested something like this... I'm not sure if he'd already had other use cases in mind. (This also seems similar/related to https://github.com/kubernetes/enhancements/issues/3784 / https://github.com/kubernetes/enhancements/pull/3787.)

pohly commented on staging/src/k8s.io/constants/annotations/annotations.go:
```diff
@@ -0,0 +1,105 @@
+	// ObjectTTLAnnotationKey represents a suggestion for kubelet for how long
+	// it can cache an object (e.g. secret, config map) before fetching it
+	// again from apiserver. This annotation can be attached to node.
+	ObjectTTLAnnotationKey = "node.alpha.kubernetes.io/ttl"
```
As long as an alpha API exists, the corresponding constants exist and may be used by someone. We might as well publish them.

We might use the opportunity to document constants the same way as fields ("This is an alpha field and requires enabling the XYZ feature gate." => "This is an alpha constant and requires enabling the XYZ feature gate."), but I wouldn't block the PR by that.

dims replied:
Kept them, added a feature-gate note on PodDeletionCost.

dims replied:
Yes ("Document constants like fields")

pohly commented:
Yes, I know :sweat_smile:

But there really doesn't seem to be a good place to put it. It cannot go into `k8s.io/constants` (staging repo!). `k8s.io/utils`? Nope, depends on klog and also a bit large. We would need a new `k8s.io/types` repo which starts with not much more than `type UID string`.

dims commented on staging/src/k8s.io/constants/annotations/annotations.go:
```diff
@@ -0,0 +1,105 @@
+	// DeprecatedAppArmorBetaContainerAnnotationKeyPrefix is the prefix to an
+	// annotation key specifying a container's apparmor profile.
+	// Deprecated: use a pod or container security context `appArmorProfile` field instead.
+	DeprecatedAppArmorBetaContainerAnnotationKeyPrefix = "container.apparmor.security.beta.kubernetes.io/"
```
Fixed; same convention applied to deprecated labels.

dims commented on staging/src/k8s.io/constants/annotations/annotations.go:
```diff
@@ -0,0 +1,105 @@
+
+	// AnnotationTopologyMode can be used to enable or disable Topology Aware
+	// Routing for a Service. Well known values are "Auto" and "Disabled".
+	AnnotationTopologyMode = "service.kubernetes.io/topology-mode"
```
Done; same pass applied to labels/ and taints/. Package doc notes the convention.

dims commented on staging/src/k8s.io/constants/labels/labels.go:
```diff
@@ -0,0 +1,91 @@
+// Well-known node labels used for topology and scheduling.
+const (
+	// LabelHostname is the label key for the node hostname.
+	LabelHostname = "kubernetes.io/hostname"
```
Dropped Label prefix across the file.

dims commented on staging/src/k8s.io/constants/labels/labels.go:
```diff
@@ -0,0 +1,91 @@
+
+	// LabelMetadataName is the label used to automatically label namespaces
+	// with their name, so they can be selected easily by tools.
+	LabelMetadataName = "kubernetes.io/metadata.name"
```
Fixed.

dims commented on staging/src/k8s.io/constants/rfc/dns.go:
```diff
@@ -0,0 +1,50 @@
+	//
+	// This is used for service names when they need to be valid DNS names
+	// (as opposed to DNS subdomains).
+	DNS1035LabelMaxLength int = 63
```
Docs rewritten; KEP-5311 linked. Held off renaming identifiers since they're load-bearing across the tree — happy to do the rename as a follow-up PR.

dims committed:
Add k8s.io/constants staging module

New zero-dependency staging module consolidating well-known Kubernetes
constants: rfc/ (name-length limits), limits/ (label/field size limits),
labels/, annotations/, taints/.

Existing constants in k8s.io/apimachinery and k8s.io/api/core/v1 are
re-exported from the new module for backwards compatibility, and
k8s.io/api/resource now uses rfc.DNS1123SubdomainMaxLength directly
instead of pulling in apimachinery/util/validation for a single integer.
kubectl and apiserver consumers updated to use the new module where
they were previously importing apimachinery/util/validation solely for
a constant.

Identifiers in labels/annotations/taints drop the package-name prefix
(labels.Hostname, not labels.LabelHostname). Deprecated constants
carry a Deprecated* prefix and a godoc Deprecated: line.

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

pohly commented:
~~May I bring up the `UID` type again? Would that fit into `k8s.io/constants/types`?~~

No, as mentioned earlier, external packages used by Kubernetes (like k8s.io/klog/v2 or k8s.io/utils) can't depend on staging repos.

dims commented:
> No, as mentioned earlier, external packages used by Kubernetes (like k8s.io/klog/v2 or k8s.io/utils) can't depend on staging repos.

💯 recipe for disaster!

liggitt commented:
is this more likely than our other staging repos to attract use by other dependencies k/k has, causing problematic k/k → third-party → staging/src/k8s.io/constants loops?

I'm wondering if making this an actual standalone repo would be better ... it would be less convenient, certainly, but I'm not sure how frequently we expect this to change

thockin commented:
I'm with Jordan - why not just make a standalone repo?

I think some of these things (e.g. protocol strings) are obviously better shared.  Some of them (e.g. format lengths) are weird to be be so removed from the rest of their validation.

thockin commented on staging/src/k8s.io/api/resource/v1/types.go:
```diff
@@ -291,7 +291,7 @@ const ResourceSliceMaxDevices = 128
 const ResourceSliceMaxDevicesWithAdvancedFeatures = 64
 
-const PoolNameMaxLength = validation.DNS1123SubdomainMaxLength // Same as for a single node name.
+const PoolNameMaxLength = rfc.DNS1123SubdomainMaxLength // Same as for a single node name.
```
As declarative validation picks up steam, these consts should not be needed.  Instead, fields should just be decorated with a "format" which almost always encapsulates a length and syntax.

My bigger fear is that we move PARTS of validation logic into this generic place but not ALL of validation (it's not just constants).  I'd rather leave these things out of this PR, let DV mature and propagate, and see if any of these things are left at the end.

BenTheElder replied:
This sort of constant could be useful for purposes other than validation? What if I'm writing a tool that wants to generate valid values without my own magic consts?

thockin replied:
Then you need more than the length, you need to format and a regex isn't going to help.

dims commented:
> I'm with Jordan - why not just make a standalone repo?

we can for sure. that's the other alternative! :)

thockin commented on staging/src/k8s.io/constants/labels/labels.go:
```diff
@@ -0,0 +1,102 @@
+	TopologyRegion = "topology.kubernetes.io/region"
+
+	// OSStable is the label key for the node operating system.
+	OSStable = "kubernetes.io/os"
```
Why "Stable" ?

dims replied:
`LabelOSStable   = "kubernetes.io/os"` is what exists - https://github.com/kubernetes/kubernetes/blob/f82d2d6ae29b6dd8e6f09066060802c2dc611b2f/staging/src/k8s.io/api/core/v1/well_known_labels.go#L45

happy to fix! suggestions please

thockin commented on staging/src/k8s.io/constants/labels/labels.go:
```diff
@@ -0,0 +1,102 @@
+// "labels.LabelHostname".
+package labels
+
+// Node labels used for topology and scheduling.
```
The problem with commenting these as "node labels" (etc) is that it is non-exclusive.  Some of these labels are also copied to pods and can be used in other places where (e.g.) region/zone matters.

jpbetz commented on staging/src/k8s.io/constants/README.md:
```diff
@@ -0,0 +1,146 @@
+
+## Packages
+
+- `rfc/` — length constants for Kubernetes name-like fields (historically named after DNS 1123/1035)
```
```suggestion
- `formats/` — constants for string field formats used in the Kubernetes API such as the name-like fields (historically named after DNS 1123/1035)
```

?

I'm not entirely convinced that moving just the length limits of the string formats results in a clean layering of modules.  Providing the format lengths but not the regex feels piecemeal. But I also realize that if we also move the regex's that we add the regex dependency we're trying to avoid..

BenTheElder replied:
regex would be OK in its own *package* IMHO, because stdlib deps don't affect the dependency graph unless we're talking about language level (the go statement and new stdlib features etc)

It should be avoided in common *packages* though because a binary may not otherwise use regex and can avoid compiling it.

thockin replied:
Most of validation is not implemented via regex, so publishing regexes would  just be a recipe for drift.  They are incomprehensible in practice.

thockin commented on staging/src/k8s.io/constants/labels/labels.go:
```diff
@@ -0,0 +1,102 @@
+
+// Label namespace suffixes for kubelet self-labeling.
+const (
+	// NamespaceSuffixKubelet is an allowed label namespace suffix
```
Using the word namespace here is SUPER confusing.

It's a prefix suffix, duh!  Perhaps we should switch to "Domain" or "Zone"?  `DomainKubelet` seems sufficient, and documented as covering all sub-domains of it.

dims replied:
It's from here:
https://github.com/kubernetes/kubernetes/blob/f82d2d6ae29b6dd8e6f09066060802c2dc611b2f/staging/src/k8s.io/api/core/v1/well_known_labels.go#L52

thockin commented on staging/src/k8s.io/constants/limits/limits.go:
```diff
@@ -0,0 +1,38 @@
+
+	// FieldManagerMaxLength is the maximum length of a field manager name
+	// in server-side apply operations.
+	FieldManagerMaxLength int = 128
```
This does not appear to be used in any API definition or validation outside of metav1, so I don't think we should expose it.

We can always ADD to this list, but taking away is hard.

jpbetz commented on staging/src/k8s.io/constants/rfc/dns.go:
```diff
@@ -0,0 +1,60 @@
+	// Kubernetes requires such values to consist of lower case alphanumeric
+	// characters or '-', start with an alphabetic character, and end with an
+	// alphanumeric character.
+	DNS1123LabelMaxLength int = 63
```
In https://github.com/kubernetes/kube-openapi/pull/384 we agreed to use `k8s-short-name` as the API visible identifier (in OpenAPI schemas and such) for this format. We should consider building on that naming decision here.

cc @thockin @liggitt

jpbetz commented on staging/src/k8s.io/constants/rfc/dns.go:
```diff
@@ -0,0 +1,60 @@
+	// Kubernetes requires such values to consist of lower case alphanumeric
+	// characters, '-' or '.', and to start and end with an alphanumeric
+	// character.
+	DNS1123SubdomainMaxLength int = 253
```
In https://github.com/kubernetes/kube-openapi/pull/384 we agreed to use `k8s-long-name` as the API visible identifier (in OpenAPI schemas and such) for this format. We should consider building on that naming decision here.

cc @thockin @liggitt

neolit123 commented:
+1 for standalone.
more trouble to own, i guess, but this is in the same boat as k/utils.

EDIT: but i also like the "make copies" idea.

aojea commented:
independently on the larger discussion going on the mailing list, a "point in time" effort like this will not be sustainable untilw we have some automation to enforce that any new constant is declared only in this new package

thockin commented:
I would argue that this is not for ALL constants, but only those which really span API types and apply to the whole system. 

Lots of API resources use things like "protocol", And there's no reason for them to have different strings (or worse, non-strings) which all mean "TCP", especially when concepts are translated or linked between those different APIs. 

But there are plenty of APIs that have their own constants that are only used in those API groups. Those don't need to be in the project-wide common pkg, because they're not common. 

 Perhaps each API group directory should have a sub package for Constance which are shared across the versions of that group.

dims commented:
dropping this.
